### ISSUE TITLE
L2-norm problem

### ISSUE THREAD

**Issue by beyondpzk** [facebookresearch/barlowtwins#2]
I have seen the l2-norm operator in the paper, ie. the formula (2), but I have not seen it in your codes, please tell me why? Thanks!

**Comment by jzbontar** [comment]
The normalization in formula (2) is performed by the calls to `self.bn`, the batch norm operator.

https://github.com/facebookresearch/barlowtwins/blob/main/main.py#L204

**Comment by MacroMayhem** [comment]
@jzbontar Please do correct me if I am wrong. Won't the [batch-norm](https://pytorch.org/docs/stable/generated/torch.nn.BatchNorm1d.html) only ensure that the features are mean-centered which corresponds to the pseudo-code (norm along the batch dimension)? I still don't understand where the L2 norm of equation(2) comes in. Thanks.

**Comment by jzbontar** [comment]
Let's start with equation (2) in the paper. In the equation we divide by the l2-norms of vectors zA and zB. Because the two vectors are zero-mean (which we mention in the paper) the l2 norms of the vectors will be the same as their standard deviations * sqrt(1 / n), where n is the size of the vector. 

In the pseudo-code and the code there is no division by the l2 norm, but there is a division by the standard deviation, which has the same effect. I hope that's clear now.
